ui/gtk3: Append non-glyph characters at last order for partial annotations

Manish7093 opened the pull request:
`BUG=rhbz#2267613`
Appended non-glyph characters at last order for partial annotations in emojier listing
- Added `ECheckVisibleLabel` class to check if an emoji can be rendered properly
- Added `non_glyph_emojis` list to track non-glyph characters

mike-fabian commented on ui/gtk3/emojier.vala:
```diff
@@ -1242,6 +1242,44 @@ public class IBusEmojier : Gtk.ApplicationWindow {
+        var pango_context = label.get_pango_context();
+        bool is_emoji_supported(string emoji) {
+            var layout = new Pango.Layout(pango_context);
+            var font_desc = Pango.FontDescription.from_string("Noto Color Emoji");
```
What happens if "Noto Color Emoji" is not the default emoji font? Is it OK to hardcode "Noto Color Emoji" here? 

Maybe use 'emoji' instead?:

mfabian@f41:~$ fc-match emoji
NotoColorEmoji.ttf: "Noto Color Emoji" "Regular"
mfabian@f41:~$

fujiwarat replied:
`IBusEmojier.EWhiteLabel` includes the custom font `m_emoji_font_family`.

fujiwarat commented:
I don't think to hide the non-glyph chars. Please append those to the last order instead and change the subject.

Manish7093 commented:
Now non-glyph characters are appended at last
![Pasted image](https://github.com/user-attachments/assets/346b4c6b-1422-4958-97f4-9d6c24605c69)

fujiwarat commented:
If you update the patch, please rebase your repo because the latest main repo fixes the CI error:
https://github.com/ibus/ibus/actions/runs/12830799599/job/36342346172?pr=2712

Also `BUG=rhbz#2267613` would be good in the patch description.

mike-fabian commented:
> Are you able to use Pango without GTK to check the non-glyph chars? Unfortunately there is a regression with your patch. Since it calls Gtk.Label for all emoji candidates, it takes a time lag to show the candidate window which depends on the number of emojis. I guess the font check does not require the display? You might like to create a COPR repo in Fedora to ask the feedback of your internal staffs.

As I am using a similar check in my emoji picker, I tried whether I could make it work without using a Gtk. I was not able to make it work without any Gtk widgets. I found that I could use a more lightweight widget, Gtk.DrawingArea(), but the speedup was insignificant (< 1%). Probably not worth it, maybe rendering works slightly different in Gtk.DrawingArea and as I am interested how my program shows Emoji in Gtk.Label, I thought it was safer to keep using Gtk.Label.

Then I tried to create a Gtk.Label only once and reuse it. This gave me a speedup of about 25%: 
https://github.com/mike-fabian/ibus-typing-booster/commit/675012ef54f22b82245a22a6d4b49f3ccff9d1a2

mike-fabian commented:
> > Are you able to use Pango without GTK to check the non-glyph chars? Unfortunately there is a regression with your patch. Since it calls Gtk.Label for all emoji candidates, it takes a time lag to show the candidate window which depends on the number of emojis. I guess the font check does not require the display? You might like to create a COPR repo in Fedora to ask the feedback of your internal staffs.
> 
> As I am using a similar check in my emoji picker, I tried whether I could make it work without using Gtk. I was not able to make it work without any Gtk widgets. I found that I could use a more lightweight widget, Gtk.DrawingArea(), but the speedup was insignificant (< 1%). Probably not worth it, maybe rendering works slightly different in Gtk.DrawingArea and as I am interested how my program shows Emoji in Gtk.Label, I thought it was safer to keep using Gtk.Label.
> 
> Then I tried to create a Gtk.Label only once and reuse it. This gave me a speedup of about 25%: [mike-fabian/ibus-typing-booster@675012e](https://github.com/mike-fabian/ibus-typing-booster/commit/675012ef54f22b82245a22a6d4b49f3ccff9d1a2)

Actually I could get another performance improvement by also reusing the Pango.Layout. Together with the reuse of the Gtk.Label this made my font check about 33% faster:

https://github.com/mike-fabian/ibus-typing-booster/commit/48cbf486bca0f105c6dd053ce16bfcadb17ca3bb

fujiwarat commented:
If you cannot delete GTK to check the glyphs, another idea is to run the check for the active page only in ibus/ui/gtk3/emojier:m_lookup_table.

E.g. If the first page in 19 pages is shown, move the chars of no glyph of the 10 candidates to the last of the candidates and shift chars with glyphs of the next candidates to the 10 candidates.
If the last page in 19 pages is shown, move the chars with glyphs to the candidates of 18th page and shift chars of no glyphs of the 18th page to the last.
The idea might need to update APIs of IBusLookupTable to be able to reorder the candidates.

fujiwarat commented on ui/gtk3/emojier.vala:
```diff
@@ -1239,9 +1239,47 @@ public class IBusEmojier : Gtk.ApplicationWindow {
+        var layout = new Pango.Layout(pango_context);
+        var font_desc = Pango.FontDescription.from_string(m_emoji_font_family);
+        layout.set_font_description(font_desc);
+        bool is_emoji_supported(string emoji) {
```
Please rename `is_emoji_supported()` to `is_visible()` and move the inner function to an API of the new class likes `emojier.ECheckVisibleLabel.is_visible()`.

fujiwarat commented on ui/gtk3/emojier.vala:
```diff
@@ -1271,7 +1309,11 @@ public class IBusEmojier : Gtk.ApplicationWindow {
                 foreach (unowned string emoji in sub_emojis) {
                     if (total_emojis.find_custom(emoji, GLib.strcmp) == null) {
-                        sorted_emojis.insert_sorted(emoji, GLib.strcmp);
+                        if (is_emoji_supported(emoji)) {
```
s/is_emoji_supported(emoji)/label.is_visible(emoji)/.

fujiwarat replied:
```
                if (label.is_visible(emoji)) {
                    total_emojis.append(emoji);
                } else if (non_glyph_emojis.find_custom(emoji, GLib.strcmp) == null) {
                    non_glyph_emojis.append(emoji);
                }
```

The similar codes exists in several lines so it's good to have an inner function in `emojier.lookup_emojis_from_annotation()` likes:

```
        void check_if_non_glyph_emojis(string emoji) {
            if (label.is_visible(emoji)) {
                total_emojis.append(emoji);
            } else if (non_glyph_emojis.find_custom(emoji, GLib.strcmp) == null) {
                non_glyph_emojis.append(emoji);
            }
        }
```

Manish7093 replied:
Created a  inner function `check_if_non_glyph_emojis()` inside `emojier.lookup_emojis_from_annotation()` function

fujiwarat commented on ui/gtk3/emojier.vala:
```diff
@@ -1282,29 +1324,39 @@ public class IBusEmojier : Gtk.ApplicationWindow {
-            foreach (unowned string emoji in sub_emojis)
-                total_emojis.append(emoji);
+            foreach (unowned string emoji in sub_emojis) {
+                if (is_emoji_supported(emoji)) {
```
Ditto.

Manish7093 replied:
Replaced with `check_if_non_glyph_emojis()`

fujiwarat commented on ui/gtk3/emojier.vala:
```diff
@@ -1282,29 +1324,39 @@ public class IBusEmojier : Gtk.ApplicationWindow {
             string ch = code.to_string();
             if (total_emojis.find_custom(ch, GLib.strcmp) == null) {
-                total_emojis.append(ch);
+                if (is_emoji_supported(ch)) {
```
Ditto.

Manish7093 replied:
Replaced with `check_if_non_glyph_emojis()`

fujiwarat commented on ui/gtk3/emojier.vala:
```diff
@@ -1282,29 +1324,39 @@ public class IBusEmojier : Gtk.ApplicationWindow {
+                    foreach (unichar code in sub_unicodes) {
+                        string ch = code.to_string();
+                        if (sorted_unicodes.find_custom(ch, GLib.strcmp) == null) {
+                            if (is_emoji_supported(ch)) {
```
Ditto.

Manish7093 replied:
Replaced with `check_if_non_glyph_emojis()`

fujiwarat commented on ui/gtk3/emojier.vala:
```diff
@@ -1239,9 +1239,47 @@ public class IBusEmojier : Gtk.ApplicationWindow {
+        var font_desc = Pango.FontDescription.from_string(m_emoji_font_family);
+        layout.set_font_description(font_desc);
+        bool is_emoji_supported(string emoji) {
+            // Remove variation selectors as they don't affect support
```
s/support/to visible/

Manish7093 replied:
Renamed is_emoji_supported() to is_visible() and moved the inner function to an API of the new class emojier.ECheckVisibleLabel.is_visible()

fujiwarat commented on ui/gtk3/emojier.vala:
```diff
@@ -1239,9 +1239,47 @@ public class IBusEmojier : Gtk.ApplicationWindow {
+            if (cleaned_emoji == "")
+                return false;
+            layout.set_text(cleaned_emoji, -1);
+            // Get the first line
```
Please delete the comment since the comment just explains the API of `layout.get_line_readonly()`.

Manish7093 replied:
Deleted the comment

fujiwarat commented on ui/gtk3/emojier.vala:
```diff
@@ -1239,9 +1239,47 @@ public class IBusEmojier : Gtk.ApplicationWindow {
+            unowned Pango.LayoutLine? line = layout.get_line_readonly(0);
+            if (line == null)
+                return false;
+            // Check ink rectangle
```
Please delete the comment since the comment just explain `ink_rect`.

Manish7093 replied:
Deleted the comment

fujiwarat reviewed (changes requested):
I evaluate the time lag again and found it's not a regression.
Currently `Emojier.m_unicode_progress_bar` is shown in the Emojier category dialog only but not in the emoji candidate window, I.e. a dialog from `Super-period, space, space` keys.

Now I think another notification popup is needed if `Emojier.m_loaded_unicode` is false in `Emojier.lookup_emojis_from_annotation()` because `Emojier.m_name_to_unicodes_dict` has the elements after the user shows the emoji preedit of "e" at the first time to reduce the memory usage and it causes a time lag.

I thought to use libnotify at first but it does not work with the  Wayland input-method protocol V2.
Currently I'm thinking to add a new D-Bus method from ibus-exntension-gtk3 to ibus-ui-gtk3 to send an error message since ibus-ui-gtk3 manages the Wayland panel but it would be a next step after this patch.

Manish7093 commented on ui/gtk3/emojier.vala:
```diff
@@ -1239,9 +1239,47 @@ public class IBusEmojier : Gtk.ApplicationWindow {
         unowned GLib.SList<unichar>? sub_exact_unicodes = null;
         unowned GLib.SList<unichar>? sub_unicodes = null;
+        // Create Pango context for font checking
+        var label = new Gtk.Label("");
```
Created a new class `emojier.ECheckVisibleLabel` extended from `emojier.EWhiteLabel`

Manish7093 commented:
> If you update the patch, please rebase your repo because the latest main repo fixes the CI error: https://github.com/ibus/ibus/actions/runs/12830799599/job/36342346172?pr=2712
> 
Rebased the patch to the latest commit

> Also `BUG=rhbz#2267613` would be good in the patch description.
>
Added

fujiwarat reviewed (changes requested):
Found some valac warnings and some suggestions for ibus codes.

fujiwarat commented:
>> Also BUG=rhbz#2267613 would be good in the patch description.

> Added

You added the line to the pull request UI but I asked to add it to the patch instead.

fujiwarat commented:
Ping @Manish7093

Manish7093 commented on ui/gtk3/emojier.vala:
```diff
@@ -169,6 +169,47 @@ public class IBusEmojier : Gtk.ApplicationWindow {
                 set_label(text);
         }
     }
+    private class ECheckVisibleLabel : EWhiteLabel {
```
Added class comment

Manish7093 commented on ui/gtk3/emojier.vala:
```diff
@@ -169,6 +169,47 @@ public class IBusEmojier : Gtk.ApplicationWindow {
+                set_label(text);
+            var pango_context = get_pango_context();
+            layout = new Pango.Layout(pango_context);
+            var font_desc = Pango.FontDescription.from_string(m_emoji_font_family);
```
Done

Manish7093 commented on ui/gtk3/emojier.vala:
```diff
@@ -169,6 +169,47 @@ public class IBusEmojier : Gtk.ApplicationWindow {
+            var font_desc = Pango.FontDescription.from_string(m_emoji_font_family);
+            layout.set_font_description(font_desc);
+        }
+        public bool is_visible(string emoji) {
```
Renamed to is_glyph_visible

Manish7093 commented on ui/gtk3/emojier.vala:
```diff
@@ -169,6 +169,47 @@ public class IBusEmojier : Gtk.ApplicationWindow {
+            line.get_pixel_extents(out ink_rect, out logical_rect);
+            if (ink_rect.width <= 0 || ink_rect.height <= 0)
+                return false;
+            if (cleaned_emoji.char_count() == 1) {
```
Added comment

Manish7093 commented on ui/gtk3/emojier.vala:
```diff
@@ -1239,9 +1280,18 @@ public class IBusEmojier : Gtk.ApplicationWindow {
         unowned GLib.SList<unichar>? sub_exact_unicodes = null;
         unowned GLib.SList<unichar>? sub_unicodes = null;
+        var label = new ECheckVisibleLabel();
+        void check_if_non_glyph_emojis(string emoji, ref GLib.SList<string>? emojis) {
```
Done suggested change

Manish7093 commented on ui/gtk3/emojier.vala:
```diff
@@ -1282,29 +1332,27 @@ public class IBusEmojier : Gtk.ApplicationWindow {
-                total_emojis.append(emoji);
+            foreach (unowned string emoji in sub_emojis) {
+                check_if_non_glyph_emojis(emoji, ref total_emojis);
+            }
```
Deleted

Manish7093 commented on ui/gtk3/emojier.vala:
```diff
@@ -1282,29 +1332,27 @@ public class IBusEmojier : Gtk.ApplicationWindow {
             }
```
Deleted

Manish7093 commented on ui/gtk3/emojier.vala:
```diff
@@ -1282,29 +1332,27 @@ public class IBusEmojier : Gtk.ApplicationWindow {
+                    foreach (unichar code in sub_unicodes) {
+                        string ch = code.to_string();
+                        if (sorted_unicodes.find_custom(ch, GLib.strcmp) == null) {
+                            check_if_non_glyph_emojis(ch, ref sorted_unicodes);
```
Done

Manish7093 commented on ui/gtk3/emojier.vala:
```diff
@@ -1282,29 +1332,27 @@ public class IBusEmojier : Gtk.ApplicationWindow {
         sub_exact_unicodes = m_name_to_unicodes_dict.lookup(annotation);
         foreach (unichar code in sub_exact_unicodes) {
             string ch = code.to_string();
             if (total_emojis.find_custom(ch, GLib.strcmp) == null) {
```
Deleted

Manish7093 commented on ui/gtk3/emojier.vala:
```diff
@@ -1271,7 +1321,7 @@ public class IBusEmojier : Gtk.ApplicationWindow {
                 foreach (unowned string emoji in sub_emojis) {
                     if (total_emojis.find_custom(emoji, GLib.strcmp) == null) {
-                        sorted_emojis.insert_sorted(emoji, GLib.strcmp);
+                        check_if_non_glyph_emojis(emoji, ref sorted_emojis);
```
Done

fujiwarat commented on ui/gtk3/emojier.vala:
```diff
@@ -1346,6 +1413,11 @@ public class IBusEmojier : Gtk.ApplicationWindow {
                 }
             }
```
Let's delete the braces for one line.

```suggestion
            }
            foreach (string ch in sorted_unicodes) {
                if (total_emojis.find_custom(ch, GLib.strcmp) == null)
                    total_emojis.append(ch);
            }
```

Manish7093 replied:
Deleted

fujiwarat commented on ui/gtk3/emojier.vala:
```diff
@@ -1346,6 +1413,11 @@ public class IBusEmojier : Gtk.ApplicationWindow {
             }
         }
+        foreach (string emoji in non_glyph_emojis) {
+            if (total_emojis.find_custom(emoji, GLib.strcmp) == null) {
+                total_emojis.append(emoji);
+            }
```
Let's delete braces for one line.
```suggestion
            if (total_emojis.find_custom(emoji, GLib.strcmp) == null)
                total_emojis.append(emoji);
```

fujiwarat reviewed (changes requested):
Thank you for the updates and I think the patch is almost ready.

> Added in the patch description

You have to insert a null line between the subject(ui/gtk3: Append...) and contents(BUG=...) in the patch.
You could find the problem when you run `git format-patch HEAD^` for your patch.

Manish7093 commented on ui/gtk3/emojier.vala:
```diff
@@ -1268,12 +1315,29 @@ public class IBusEmojier : Gtk.ApplicationWindow {
         unowned GLib.SList<unichar>? sub_exact_unicodes = null;
         unowned GLib.SList<unichar>? sub_unicodes = null;
+        var label = new ECheckVisibleLabel();
+        CheckGlyph check_if_non_glyph_emojis = (emoji, ref emojis, do_sort) => {
```
Added Comment

Manish7093 commented on ui/gtk3/emojier.vala:
```diff
@@ -1303,40 +1367,43 @@ public class IBusEmojier : Gtk.ApplicationWindow {
-                total_emojis.append(emoji);
+               check_if_non_glyph_emojis(emoji,
+                                         ref total_emojis,
+                                         false);
```
Done

Manish7093 commented on ui/gtk3/emojier.vala:
```diff
@@ -165,6 +165,53 @@ public class IBusEmojier : Gtk.ApplicationWindow {
                 set_label(text);
         }
     }
+    /*
```
Done
